maitake/src/wait/map/tests.rs

```diff
@@ -0,0 +1,368 @@
+            scheduler.spawn(async move {
+                let val = q.wait(i).await.unwrap();
+                COMPLETED.fetch_add(1, Ordering::SeqCst);
+
+                assert_eq!(val, 100 + i);
+
+                if i < (TASKS - 1) {
+                    assert!(matches!(q.wake(&(i + 1), 100 + i + 1), WakeOutcome::Woke));
+                } else {
+                    assert!(matches!(
+                        q.wake(&(i + 1), 100 + i + 1),
+                        WakeOutcome::NoMatch(_)
+                    ));
+                }
+            });
+        }
+
+        let tick = scheduler.tick();
+
+        assert_eq!(tick.completed, 0);
+        assert_eq!(COMPLETED.load(Ordering::SeqCst), 0);
+        assert!(!tick.has_remaining);
+
+        q.wake(&0, 100);
+
+        let tick = scheduler.tick();
+
+        assert_eq!(tick.completed, TASKS);
+        assert_eq!(COMPLETED.load(Ordering::SeqCst), TASKS);
+        assert!(!tick.has_remaining);
+    }
+
+    #[derive(Debug)]
+    struct CountDropKey {
+        idx: usize,
+        cnt: &'static AtomicUsize,
+    }
+
+    impl PartialEq for CountDropKey {
+        fn eq(&self, other: &Self) -> bool {
+            self.idx.eq(&other.idx)
+        }
+    }
+
+    impl Drop for CountDropKey {
+        fn drop(&mut self) {
+            self.cnt.fetch_add(1, Ordering::SeqCst);
+        }
+    }
+
+    #[derive(Debug)]
+    struct CountDropVal {
+        cnt: &'static AtomicUsize,
+    }
+
+    impl Drop for CountDropVal {
+        fn drop(&mut self) {
+            self.cnt.fetch_add(1, Ordering::SeqCst);
+        }
+    }
+
+    #[test]
+    fn drop_no_wake() {
+        crate::util::trace_init();
+        static COMPLETED: AtomicUsize = AtomicUsize::new(0);
+        static KEY_DROPS: AtomicUsize = AtomicUsize::new(0);
+        static VAL_DROPS: AtomicUsize = AtomicUsize::new(0);
+
+        let scheduler = Scheduler::new();
+        let q = Arc::new(WaitMap::<CountDropKey, CountDropVal>::new());
+
+        const TASKS: usize = 10;
+
+        for i in 0..TASKS {
+            let q = q.clone();
+            scheduler.spawn(async move {
+                q.wait(CountDropKey {
+                    idx: i,
+                    cnt: &KEY_DROPS,
+                })
+                .await
+                .unwrap_err();
+                COMPLETED.fetch_add(1, Ordering::SeqCst);
+            });
+        }
+
+        let tick = scheduler.tick();
+
+        assert_eq!(tick.completed, 0);
+        assert_eq!(COMPLETED.load(Ordering::SeqCst), 0);
+        assert_eq!(KEY_DROPS.load(Ordering::SeqCst), 0);
+        assert_eq!(VAL_DROPS.load(Ordering::SeqCst), 0);
+        assert!(!tick.has_remaining);
+
+        q.close();
+
+        let tick = scheduler.tick();
+
+        assert_eq!(tick.completed, TASKS);
+        assert_eq!(COMPLETED.load(Ordering::SeqCst), TASKS);
+        assert_eq!(KEY_DROPS.load(Ordering::SeqCst), TASKS);
+        assert_eq!(VAL_DROPS.load(Ordering::SeqCst), 0);
```

Comment:
nit, take it or leave it: all the atomic ops in these tests *can* be `Relaxed` --- the tests are all single-threaded, so these atomics will never be modified anywhere else, so ordering doesn't matter at all. also, they're just counters, so they can always be `Relaxed`. not a big deal though.